[BC-BREAKING] Replace Type dispatch with ATenDispatch

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21320 [wip] Replace Type dispatch with ATenDispatch**

Replace Type object based dispatch with a table of void*. This is bc breaking because it will break cpp extensions that extend Type.

This gives us a registration system that has a similar API to c10 dispatch. Backend extensions can be folded into use this API directly rather than having its own dispatch system. This also saves us a dispatch because Type dispatch was doing a dynamic dispatch and a vtable dispatch, while this table only does a dynamic dispatch.

As a follow up to this change, we can clean up a lot of code involving Type.

Outline of this change:
- Create dispatcher to register functions and variable functions
- Generate Type methods as static instead so that they can be registered on dispatcher
- Generate VariableType methods as static, instead of storing baseType, Variable functions redispatch and on the redispatch, it will call the non variable method.
- Change Functions.h and TensorMethods.h to call into the dispatcher rather than LegacyTypeDispatch.
- Factory functions previously used getType to initialize backends. New dispatcher doesn't do this, so for now, make a call to legacyTypeDispatch() to initialize backends.

Benchmarks are a bit noisy because we're timing such a small op, but I ran benchmarks on a null op, with and without variable unwrapping.
```
a = torch.randn(1,1)
%timeit torch.null(a)
%timeit torch.null_with_backwards(a)
```
Before, without variable unwrapping: ~365ns
After, without variable unwrapping: ~330ns
Before, with variable unwrapping: ~450ns
After, with variable unwrapping: ~420ns

Differential Revision: [D15637222](https://our.internmc.facebook.com/intern/diff/D15637222)

Commit:
Replace Type dispatch with ATenDispatch

Commit:
Update on "Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[wip] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[wip] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[wip] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

ezyang commented:
Do you want initial review now or wait until it's not WIP?

li-roy commented:
@ezyang sorry this is ready for review, i forgot to change the title.

ezyang commented:
cc @ailzhang What's the plan for coordinating this diff with XLA?

ezyang commented:
UBSAN seems to have caught an error:

```
un 05 03:14:01 [2/2] c++ complex_registration_extension.o -shared -o complex_registration_extension.so
Jun 05 03:14:01 /var/lib/jenkins/workspace/aten/src/ATen/core/LegacyTypeDispatch.h:86:70: runtime error: reference binding to null pointer of type 'const at::Type'
Jun 05 03:14:01     #0 0x7fbcb0d36b6f in at::LegacyTypeDispatch::getTypeRaw(c10::Backend, c10::ScalarType, bool) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0xf77b6f)
Jun 05 03:14:01     #1 0x7fbcb0d3698e in at::legacyTensorType(c10::TensorImpl const&) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0xf7798e)
Jun 05 03:14:01     #2 0x7fbcb0d36783 in at::Tensor::dispatch_type() const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0xf77783)
Jun 05 03:14:01     #3 0x7fbcb0bd26aa in torch::autograd::THPVariable_type(_object*, _object*, _object*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0xe136aa)
Jun 05 03:14:01     #4 0x5565dc3a5743 in _PyCFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/methodobject.c:231
Jun 05 03:14:01     #5 0x5565dc42c42b in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4851
Jun 05 03:14:01     #6 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #7 0x5565dc42653a in _PyFunction_FastCall /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4933
Jun 05 03:14:01     #8 0x5565dc42653a in fast_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4968
Jun 05 03:14:01     #9 0x5565dc42c504 in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4872
Jun 05 03:14:01     #10 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #11 0x5565dc4258e3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #12 0x5565dc426e69 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5084
Jun 05 03:14:01     #13 0x5565dc3a5b0e in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2310
Jun 05 03:14:01     #14 0x5565dc3aa6a2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2373
Jun 05 03:14:01     #15 0x5565dc3a554d in PyObject_Call /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2261
Jun 05 03:14:01     #16 0x5565dc452a6b in do_call_core /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5120
Jun 05 03:14:01     #17 0x5565dc452a6b in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3404
Jun 05 03:14:01     #18 0x5565dc4258e3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #19 0x5565dc426c4b in _PyFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5084
Jun 05 03:14:01     #20 0x5565dc3a5b0e in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2310
Jun 05 03:14:01     #21 0x5565dc3aa6a2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2373
Jun 05 03:14:01     #22 0x5565dc3a554d in PyObject_Call /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2261
Jun 05 03:14:01     #23 0x5565dc3ffa90 in slot_tp_call /tmp/build/80754af9/python_1546130271559/work/Objects/typeobject.c:6207
Jun 05 03:14:01     #24 0x5565dc3a592a in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2331
Jun 05 03:14:01     #25 0x5565dc42c57d in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4875
Jun 05 03:14:01     #26 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #27 0x5565dc4258e3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #28 0x5565dc426e69 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5084
Jun 05 03:14:01     #29 0x5565dc3a5b0e in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2310
Jun 05 03:14:01     #30 0x5565dc3aa6a2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2373
Jun 05 03:14:01     #31 0x5565dc3a554d in PyObject_Call /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2261
Jun 05 03:14:01     #32 0x5565dc452a6b in do_call_core /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5120
Jun 05 03:14:01     #33 0x5565dc452a6b in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3404
Jun 05 03:14:01     #34 0x5565dc4258e3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #35 0x5565dc426c4b in _PyFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5084
Jun 05 03:14:01     #36 0x5565dc3a5b0e in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2310
Jun 05 03:14:01     #37 0x5565dc3aa6a2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2373
Jun 05 03:14:01     #38 0x5565dc3a554d in PyObject_Call /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2261
Jun 05 03:14:01     #39 0x5565dc3ffa90 in slot_tp_call /tmp/build/80754af9/python_1546130271559/work/Objects/typeobject.c:6207
Jun 05 03:14:01     #40 0x5565dc3a592a in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2331
Jun 05 03:14:01     #41 0x5565dc42c57d in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4875
Jun 05 03:14:01     #42 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #43 0x5565dc4258e3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #44 0x5565dc426e69 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5084
Jun 05 03:14:01     #45 0x5565dc3a5b0e in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2310
Jun 05 03:14:01     #46 0x5565dc3aa6a2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2373
Jun 05 03:14:01     #47 0x5565dc3a554d in PyObject_Call /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2261
Jun 05 03:14:01     #48 0x5565dc452a6b in do_call_core /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5120
Jun 05 03:14:01     #49 0x5565dc452a6b in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3404
Jun 05 03:14:01     #50 0x5565dc4258e3 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #51 0x5565dc426c4b in _PyFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5084
Jun 05 03:14:01     #52 0x5565dc3a5b0e in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2310
Jun 05 03:14:01     #53 0x5565dc3aa6a2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2373
Jun 05 03:14:01     #54 0x5565dc3a554d in PyObject_Call /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2261
Jun 05 03:14:01     #55 0x5565dc3ffa90 in slot_tp_call /tmp/build/80754af9/python_1546130271559/work/Objects/typeobject.c:6207
Jun 05 03:14:01     #56 0x5565dc3a592a in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2331
Jun 05 03:14:01     #57 0x5565dc42c57d in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4875
Jun 05 03:14:01     #58 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #59 0x5565dc42653a in _PyFunction_FastCall /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4933
Jun 05 03:14:01     #60 0x5565dc42653a in fast_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4968
Jun 05 03:14:01     #61 0x5565dc42c504 in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4872
Jun 05 03:14:01     #62 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #63 0x5565dc42653a in _PyFunction_FastCall /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4933
Jun 05 03:14:01     #64 0x5565dc42653a in fast_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4968
Jun 05 03:14:01     #65 0x5565dc42c504 in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4872
Jun 05 03:14:01     #66 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #67 0x5565dc425d70 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #68 0x5565dc426e69 in _PyFunction_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:5084
Jun 05 03:14:01     #69 0x5565dc3a5b0e in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2310
Jun 05 03:14:01     #70 0x5565dc3aa6a2 in _PyObject_Call_Prepend /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2373
Jun 05 03:14:01     #71 0x5565dc3a554d in PyObject_Call /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2261
Jun 05 03:14:01     #72 0x5565dc3ff0fa in slot_tp_init /tmp/build/80754af9/python_1546130271559/work/Objects/typeobject.c:6420
Jun 05 03:14:01     #73 0x5565dc42c766 in type_call /tmp/build/80754af9/python_1546130271559/work/Objects/typeobject.c:915
Jun 05 03:14:01     #74 0x5565dc3a592a in _PyObject_FastCallDict /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2331
Jun 05 03:14:01     #75 0x5565dc426a59 in _PyObject_FastCallKeywords /tmp/build/80754af9/python_1546130271559/work/Objects/abstract.c:2496
Jun 05 03:14:01     #76 0x5565dc42c57d in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4875
Jun 05 03:14:01     #77 0x5565dc452146 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3351
Jun 05 03:14:01     #78 0x5565dc3d422a in _PyFunction_FastCall /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4933
Jun 05 03:14:01     #79 0x5565dc42c504 in call_function /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4872
Jun 05 03:14:01     #80 0x5565dc451389 in _PyEval_EvalFrameDefault /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:3335
Jun 05 03:14:01     #81 0x5565dc427288 in _PyEval_EvalCodeWithName /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4166
Jun 05 03:14:01     #82 0x5565dc427288 in PyEval_EvalCodeEx /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:4187
Jun 05 03:14:01     #83 0x5565dc42801b in PyEval_EvalCode /tmp/build/80754af9/python_1546130271559/work/Python/ceval.c:731
Jun 05 03:14:01     #84 0x5565dc4aa3c3 in run_mod /tmp/build/80754af9/python_1546130271559/work/Python/pythonrun.c:1025
Jun 05 03:14:01     #85 0x5565dc4aa7c0 in PyRun_FileExFlags /tmp/build/80754af9/python_1546130271559/work/Python/pythonrun.c:978
Jun 05 03:14:01     #86 0x5565dc4aa9c2 in PyRun_SimpleFileExFlags /tmp/build/80754af9/python_1546130271559/work/Python/pythonrun.c:419
Jun 05 03:14:01     #87 0x5565dc4ae4b2 in run_file /tmp/build/80754af9/python_1546130271559/work/Modules/main.c:340
Jun 05 03:14:01     #88 0x5565dc4ae4b2 in Py_Main /tmp/build/80754af9/python_1546130271559/work/Modules/main.c:811
Jun 05 03:14:01     #89 0x5565dc37702d in main /tmp/build/80754af9/python_1546130271559/work/Programs/python.c:69
Jun 05 03:14:01     #90 0x7fbcc3d5282f in __libc_start_main /build/glibc-LK5gWL/glibc-2.23/csu/../csu/libc-start.c:291
Jun 05 03:14:01     #91 0x5565dc457e0d in _start /home/rdonnelly/mc/conda-bld/compilers_linux-64_1534865402226/work/.build/src/glibc-2.12.2/csu/../sysdeps/x86_64/elf/start.S:103
Jun 05 03:14:01 
Jun 05 03:14:01 SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /var/lib/jenkins/workspace/aten/src/ATen/core/LegacyTypeDispatch.h:86:70 in 
```

ezyang commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1261 @@
+#pragma once
```
This guy is generated right? Can we get `${generated_comment}` in the template?

ezyang commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -168,7 +168,7 @@ struct CAFFE2_API AutoNonVariableTypeMode {
  */
 inline Type& legacyTensorType(const TensorImpl& tensor) {
-  // NB: It's valid to use getTypeRaw here, because the TensorImpl
+  // NB: It's valid to use getTypeRaw here, because the TensorImpl	  return globalLegacyTypeDispatch().getType(
```
This diff looks botched

ezyang commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -103,7 +103,6 @@ class CAFFE2_API LegacyTypeDispatch {
     type_registry[static_cast<int>(b)] = std::move(t);
     detail::getVariableHooks().registerVariableTypeFor(this, b);
   }
-private:
```
Why'd this become public?

li-roy replied:
For now, I'm calling this directly from Functions.h for all factory functions.

ezyang commented on aten/src/ATen/core/TensorMethods.h:
```diff
@@ -57,1240 +58,1652 @@ inline void Tensor::set_data(Tensor new_data) {
 inline Tensor Tensor::abs() const {
-    return dispatch_type().abs(*this);
+    return globalATenDispatch().getOp<Tensor (const Tensor &)>(
+        tensorTypeIdToBackend(type_id()), is_variable(), 30, "abs")(*this);
```
The hard coded numbers here make me concerned about merge conflicts that occur when these numbers wobble. If I insert a new function, that will shift the numbers for everything allocated after it down one; if two people edit unrelated functions, this will assuredly cause merge conflicts from renumbering. (Also, are we **sure** that the number generation process here is deterministic? It'll be really bad if someone else runs the generator script and the numbering is different.)

ezyang replied:
A very most basic mitigation I can think of is to ensure the numbering only occurs in one place, so at least you only have one file to solve merge conflicts on. An even better mitigation is to ensure this file is NOT checked in (to do that, you'll have to solve the "we don't know how to codegen in xplat" problem one way or another.)

dzhulgakov replied:
Yes, we can move it out (I think usages in caffe2/core/operator.h are already ifdefed or can be ifdefed). We'd still need to figure out xplat codegen at some point for PT mobile track. But it's not an immediate blocker.

smessmer replied:
Let's go with the solution we discussed offline - a pointer to the dispatch table, looked up by string and stored in a static variable in the function so it's only looked up once. That seems to be the easiest solution.

ezyang commented on aten/src/ATen/core/Type.h:
```diff
@@ -140,422 +140,6 @@ struct CAFFE2_API Type {
     }
   }
 
-  // example
```
What's blocking removing `Type.h` entirely?

li-roy replied:
A few functions like unsafeTensorFromTH are still on Type and haven't found a new home yet.

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -320,7 +320,7 @@ Tensor VariableType::detach(const Tensor & self) const {
 }
 
-Tensor & VariableType::detach_(Tensor & self) const {
+Tensor & VariableType::detach_(Tensor & (*_op)(Tensor &), Tensor & self) {
```
This is pretty weird, especially since you never actually call `_op` here. Are you sure this is the pattern you want for "Variable-only" operations?

ezyang replied:
I guess this is fine.

ezyang commented on aten/src/ATen/templates/ATenDispatch.h:
```diff
@@ -0,0 +1,89 @@
+#include <c10/core/Backend.h>
+#include <unordered_map>
+
+// This dispatch class serves as a replacement for Type based dispatch.
```
Try not to write comments in reference to historically previous concepts which no longer exist in the codebase. New readers aren't going to know what "Type based dispatch" means! Explain it from first principles.

ezyang commented on aten/src/ATen/templates/ATenDispatch.h:
```diff
@@ -0,0 +1,89 @@
+  template<class FuncType>
+  ATenDispatch& registerOp(Backend backend, const char* schema, FuncType* fn) {
+   auto id = getSchemaId(schema);
+   function_table[static_cast<int64_t>(backend)][id] = reinterpret_cast<void*>(fn);
```
You don't need a `reinterpret_cast` to cast a pointer into a void pointer. (It's only needed the other way.)

ezyang replied:
So... no error checking, I guess?

ezyang commented on aten/src/ATen/templates/ATenDispatch.h:
```diff
@@ -0,0 +1,89 @@
+      if (function_table[static_cast<int64_t>(Backend::Undefined)][id] == nullptr) {
+        AT_ERROR("No function is registered for ", name, " on backend ", toString(backend));
+      }
+      function_table[static_cast<int64_t>(backend)][id] = function_table[static_cast<int64_t>(Backend::Undefined)][id];
```
A write in getOp?! Say it ain't so! (If you want to do a write in a "reader" function that is ostensibly thread safe, you had better argue why this is thread safe.)

ezyang replied:
Why not just return the appropriate op in these cases? It's not like we need the `== nullptr` case to be fast.

ezyang commented on aten/src/ATen/templates/ATenDispatch.h:
```diff
@@ -0,0 +1,89 @@
+
+    if (is_variable) {
+      if (wrapper_table[id] == nullptr) {
+        AT_ERROR("No autograd wrapper is registered for ", name, ". Please report a bug to PyTorch.");
```
If this is really a bug you should use `TORCH_INTERNAL_ASSERT`

ezyang commented on aten/src/ATen/templates/ATenDispatch.h:
```diff
@@ -0,0 +1,89 @@
+  int64_t getSchemaId(std::string schema) {
+    static std::unordered_map<std::string, int64_t> schema_to_id = {
+      ${schema_to_id_pairs}
+    };
```
@zdevito Do you know if a static variable like this takes a long time to compile? Also even though this isn't perf sensitive code, we should really avoid using `std::unordered_map`

ezyang commented on test/cpp_extensions/complex_registration_extension.cpp:
```diff
@@ -80,6 +81,9 @@ TypeID ComplexCPUType::ID() const {
 }
 
+static auto complex_empty_registration = globalATenDispatch()
+  .registerOp(Backend::ComplexCPU, "aten::empty(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor", &ComplexCPUType::empty);
```
If complex is really going to use `registerOp`, that makes me a lot more nervous about lack of error checking. (Maybe this should have been rewritten to use your other extension API?)

ezyang commented:
Biggest concern, I think, is that we should stop checking in `Tensor.h` prior to landing this. I think the original motivation for checking it in has gone away now, so we can stop checking in this autogenerated file.

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

ailzhang commented:
cc: @dlibenzi @asuhan

zdevito commented:
I didn't look closely, but generally looks good to me. So when others are good with it, I am good as well.

dzhulgakov commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1263 @@
+// This dispatch class serves as a replacement for Type based dispatch.
+// It stores a function for each native function and backend pair.
+// Functions that are the same for all backends can be registered under
+// Backend::Undefined.
```
can't you handle it at the write path? I.e. in registerOp just go and populate every entry (but skipping override)? that way can avoid logic (and mutation!) in getOp

dzhulgakov commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1263 @@
+  template<class FuncType>
+  ATenDispatch& registerOp(Backend backend, const char* schema, FuncType* fn) {
+   auto id = getSchemaId(schema);
+   function_table[static_cast<int64_t>(backend)][id] = reinterpret_cast<void*>(fn);
```
maybe at least basic error checking that we don't register the entry twice?

dzhulgakov commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1263 @@
+  ATenOp<FuncType> getOp(Backend backend, bool is_variable, int64_t id, const std::string& name) {
+    if (function_table[static_cast<int64_t>(backend)][id] == nullptr) {
+      if (function_table[static_cast<int64_t>(Backend::Undefined)][id] == nullptr) {
+        AT_ERROR("No function is registered for ", name, " on backend ", toString(backend));
```
the only place you use name is error messages. Let's not pass it in (less codegen and parameters) and instead we can use the reverse of getSchemaId to look up the name

dzhulgakov commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1263 @@
+  }
+
+  template<class FuncType>
+  ATenOp<FuncType> getOp(Backend backend, bool is_variable, int64_t id, const std::string& name) {
```
also, depending on compiler, if things don't get inlined, this can trigger memory allocation for std::string as you pass a literal from outside. string_view or char* are preferred in these cases (but you probably don't need name at all - see below

dzhulgakov commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1263 @@
+
+    if (is_variable) {
+      if (wrapper_table[id] == nullptr) {
+        AT_ERROR("No autograd wrapper is registered for ", name, ". Please report a bug to PyTorch.");
```
since the outer function is inlined - I'd pull out the error handling into a non-inlined function to reduce code size a bit (benchmarks will tell)

dzhulgakov commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1263 @@
+
+ private:
+  int64_t getSchemaId(std::string schema) {
+    static std::unordered_map<std::string, int64_t> schema_to_id = {
```
this absolutely has to be in .cpp file. First, static variables in header files is a recipe for disaster (because of ODR rules and dynamic linking), second, it increases build time for every file that includes ATenDispatch.h which is pretty much everything

smessmer commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,1261 @@
+#include <c10/core/Backend.h>
+#include <unordered_map>
+
+// This dispatch class serves as a replacement for Type based dispatch.
```
Can we mention here that this class is only an intermediate state and meant to be deleted when we move to c10?

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

dlibenzi commented:
I am curious how XLA type override is going to work with this PR in.
This totally break our model and/or forces (assuming still possible) large rewrites.
Type/TypeDefault functions are no more virtual.

soumith commented:
@li-roy by BC-Breaking, do you mean this will break everyone's cpp-extensions???

li-roy commented:
@soumith This breaks cpp extensions that override Type. That's going to be things like complex and xla.

soumith commented:
okay, got it. let's work with xla to land this, esp. around testing -- from what you said there's no end-to-end testing for these changes other than via those repos

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

ezyang commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,102 @@
+class CAFFE2_API ATenOpTable {
+ public:
+  ATenOpTable(std::string schema)
+    : schema_(schema) {}
```
nit: `std::move(schema)`

ezyang commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,102 @@
+  }
+ private:
+  void registerOp(Backend backend, void* fn) {
+    TORCH_CHECK(function_table_[static_cast<int64_t>(backend)] == nullptr,
```
Internal asserts here? (I guess if you're going to expose directly to extensions an internal assert is not appropriate, but for internal use, this isn't a public API right)

ezyang replied:
There's no type checking in the registration API right now; as long as we don't have any type checking I have a pretty strong preference of NOT allowing external parties to poke this API. I don't know what the situation is with XLA and this diff at the moment.

ezyang commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -0,0 +1,102 @@
+  void* getBaseOp(Backend backend) const {
+    if (function_table_[static_cast<int64_t>(backend)] == nullptr) {
+      TORCH_CHECK(function_table_[static_cast<int64_t>(Backend::Undefined)] != nullptr,
+          "No function is registered for schema ", schema_, " on backend ", toString(backend));
```
Re error message: not only is no function not registered, but there is no undefined implementation registered either.

Actually, why are we dispatching to `Undefined` at all, at this point? Can't we just error out straight up at this point?

ezyang commented on aten/src/ATen/core/ATenDispatch.cpp:
```diff
@@ -0,0 +1,10 @@
+
+namespace at {
+
+ATenDispatch & globalATenDispatch() {
```
Registering to the global table is not thread safe, right? Gotta be careful: library loads can happen in different threads.

ezyang reviewed (approved):
I'm not going to block landing this on account of lack of type checking, but if XLA is going to use this interface I would be veeeeery careful. But I don't know what the XLA plan is at the moment.

dlibenzi commented:
The current path to XLA is that there is an include file generated, which provides a matching between C++ signature, and the magic-string pytorch generates for table based registrations.
We then match our C++ signature with the C++ one in that file.
As long as pytorch does not allow us to register string signatures which are not known to it, it should be fine.

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head

Commit:
Update on "[BC-BREAKING] Replace Type dispatch with ATenDispatch"

Replace Type dispatch with ATenDispatch

gh-metadata: pytorch pytorch 21320 gh/li-roy/25/head
